### PR TITLE
Feat(GCP Connector Registry Deployments): Upload low-code `manifest.yml` and `python_components.zip` to GCS connector registry

### PR DESCRIPTION
**Pull request by aaronsteers**
## What

This PR adds three files to the uploads for our Low-Code connectors. All of these are uploaded as their 'latest' version as well as with their current version - which then becomes a pinnable resource.

1. `manifest.yml` - Uploading this to the GCS bucket allows us to easily download them on-demand.
   - For connectors which do not require a `components.py` file, this is all that is needed to be downloaded.
2. `python_components.zip` - If the connector requires a `components.py` file, this zip will contain both the `manifest.yml` file as well as the `components.py` file.
   - Including both files together allows this zip to then be leveraged as a fully self-contained definition of the connector.
3. `python_components.zip.sha256` - A sha256 hash of the contents of the `python_components.zip` file.
   - This serves as a checksum for consumers to verify the zip has not been tampered with. 

## How

- We add new checks and upload steps to the existing `gcs_upload.py` file.
- Where it was not otherwise possible to keep logic DRY, we add a new `_upload_file()` function which does the necessary uploading of both the versioned and latest copy of any file.

## Other Notes

- Virus checkers and browsers may prohibit downloading bare `.py` files. A common solution is to first zip the contents and secondly, provide a sha or md5 for validating that the contents have not been disturbed.
- The goal of this was just to add the `manifest.yml` file, but I decided we should just go ahead and do the python components while we are here.
- This also adds a flag to the registry so we can scan to tell if a lowcode connector needs Python or not. (`pythonComponents.required=True`).
- All low-code connectors will be xcopy-able after this PR merged.
  - Low-code connectors that don't need custom python components will be executable by downloading their `manifest.yml` file.
  - Low-code connectors that do need custom python components will be exeutable by downloading their `python_components.zip` file, which _also_ contains the `manifest.yml` file.
  - The inclusion of the yaml file in the zip is intentional, since it allows that single file to fully-define the connector.
  - I expect we'll update the CDK in the future to be able to accept the yaml file _or_ the zip file.

## Testing
You can use these commands to test uploads to dev-metadata-service bucket

```sh
echo "uploading regular connector"
poetry run metadata_service upload \
    /Users/ben/Development/repos/airbyte/airbyte-integrations/connectors/source-s3/metadata.yaml \
    /Users/ben/Development/repos/airbyte/docs \
    dev-airbyte-cloud-connector-metadata-service

echo "uploading prerelease"
poetry run metadata_service upload \
    /Users/ben/Development/repos/airbyte/airbyte-integrations/connectors/source-s3/metadata.yaml \
    /Users/ben/Development/repos/airbyte/docs \
    dev-airbyte-cloud-connector-metadata-service \
    --prerelease 1.0.0-beta \
    --disable-dockerhub-checks

echo "uploading manifest only"
poetry run metadata_service upload \
    /Users/ben/Development/repos/airbyte/airbyte-integrations/connectors/source-aha/metadata.yaml \
    /Users/ben/Development/repos/airbyte/docs \
    dev-airbyte-cloud-connector-metadata-service

echo "uploading manifest + components.py"
poetry run metadata_service upload \
    /Users/ben/Development/repos/airbyte/airbyte-integrations/connectors/source-the-guardian-api \
    /Users/ben/Development/repos/airbyte/docs \
    dev-airbyte-cloud-connector-metadata-service
```

### CODE REVIEW

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (aaronsteers) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:256]: This `md5` would be a minimum checksum for now. During execution, the caller would check the `components.py` contents against the hash before executing.

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (aaronsteers) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:350]: I'm not sure what "_doc_upload()" means - if "doc" means any document or a specific documentation artifact. Rather than wrapping in a new custom function, I just used the lower-level "upload_file_if_changed()" function that other functions seem to be wrapping.

Comment (aaronsteers) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:350]: TODO: probably we want to zip or pack this before uploading. I think otherwise, antivirus programs may be tempted to delete/quarantine the file when it is downloaded.

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (aaronsteers) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:210]:
```suggestion
        upload_as_version: The version to upload the file as or 'False' to skip uploading
            the versioned copy.
```

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (aaronsteers) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:301]:
```suggestion
    """Apply the components.py file data to the metadata file before uploading it to GCS.

    Currently, we only set the boolean `required=True` and we provide a hash of the
    expected contents of `components.py`.
    """
    if components_py_file_path:
```

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (aaronsteers) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:426]:
```suggestion
        local_path=python_components_zip_sha256_file_path, gcp_connector_dir=gcp_connector_dir, bucket=bucket,
```

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (aaronsteers) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:498]:
```suggestion
                description="latest manifest.yml file",
```

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (aaronsteers) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:480]:
```suggestion
                description="versioned manifest.yml file",
```

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (aaronsteers) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:4]:
```suggestion
```

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (aaronsteers) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:461]:
```suggestion
```

**Comment by aaronsteers** [comment]
@bnchrch - Can you take a look at this PR and see if the approach looks good? And any tips on how to test?

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (natikgadzhi) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:316]: To have that sha in metadata, you'd need to also change the metadata schema. @alafanechere pointed me at where it is, but of course I already forgot.

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (natikgadzhi) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:198]: Nit: should this also have a default?

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (natikgadzhi) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:192]: I wonder if it would be easier to have `_file_upload` upload once, with a tag that can be either a string version, or latest. That could return a named `@dataclass` instead of a tuple just for readability. And we could call this for both latest and version number when required.

This is purely stylistic / readability, so feel free to ignore me on this.

**Review by natikgadzhi — COMMENTED** [review]
I'm not the most up to speed person on this, left a few comments.

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (natikgadzhi) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:316]: Found it: https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/metadata_service/lib/metadata_service/models/src/ConnectorTestSuiteOptions.yaml#L10

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (bnchrch) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:199]: 💅 Might be high time to formalize that tuple
```suggestion
) -> tuple[MaybeUpload, MaybeUpload]:
```

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (bnchrch) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:395]: We will need to update airbyte_ci Metadata step as well to include this file https://github.com/airbytehq/airbyte/blob/07-30-feat-cat_add_manifest_support/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/metadata/pipeline.py/#L82-L85

**Review by bnchrch — APPROVED** [review]
This is awesome! AJ, thank you!

Most things are small style asks with the one exception being we need to also update the `aibryte_ci` container.

Though that could be pushed to another PR, so Im happy to pre-approve.

**On the note of testing:** 

I think we would benefit with a  test case for
* `does _file_upload get called when components.py is present`

To do that you could either make it its own test case or extend `test_upload_metadata_to_gcs_valid_metadata` to:
* Have a paramaterized field for `has_components_py_file`
* Assert if `_file_upload` is called with the expected arguments (or not called)


Also you could add test cases for:
* `compute_sha256`
* `_file_upload`

These would be useful, but optional

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (aaronsteers) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:197]: Actually, neither have a default right now - this is to ensure an arg isn't forgotten. Since this is code we control and only is for us, I sacrificed some convenience in favor of keeping things explicit.

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (aaronsteers) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:191]: I appreciate this feedback but without larger refactoring effort, I think this is the most streamlined option we have. I started with the approach you recommend, actually, but because we redundantly have to pass the bucket and several other args, it was much more verbose in that form - and then the "if/then" logic was also having to be repeated in several places.

One thing I can do to improve readability without a lot of refactoring would be to use the Named tuple or type alias approach that @bnchrch suggested in another thread. 

https://github.com/airbytehq/airbyte/pull/42584#discussion_r1697339609

@bnchrch suggests "MaybeUpload" as a pairing of the bool/str pair, which I think simplifies the return signature significantly. We could do other refactoring - such as using the null string as a replacement for a falsey bool - but I'm reluctant to do such a large refactor here where we're just trying to add more files to the process.

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (aaronsteers) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:395]: @bnchrch - I don't think I follow this comment.

Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:395]: @aaronsteers Sorry! The link is broken
https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/metadata/pipeline.py/#L82-L85

Essentially we'll have to mount the components file to the metadata upload container. 

Actually lets just mount the whole connector.

Ill make that change for you!

Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:395]: @aaronsteers Just pushed a commit with the change.

cc: @alafanechere for your thoughts on this as the tooling wizard.

Essentially instead of always updating the ci steps to include more and more individual files for validation, I just mounted the entire connector.

This will come with a publish speed penalty.

But perhaps if that speed penalty becomes a penalty we look at 
1. mounting the connector once
2. using a shared volume to prevent re-mounts?

I dont know if thats possible in dagger yet though

**Comment by bnchrch** [comment]
@aaronsteers @erohmensing Just a heads up that Ive taken over this PR as of friday.

This morning I pushed up some WIP changes to refactor gcs_upload to use AJ's `_file_upload` method.

Need to
1. update the test
2. test that that this works.

Just transitioning to an OC issue right now

**Comment by bnchrch** [comment]
@erohmensing This PR is ready for review!

(@aaronsteers thanks for letting me take this over)

Since the last review you'll find
1. Large refactors to use AJ's file_upload command
2. Heavily Update unit tests
3. A new flag to disable docker checks

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (erohmensing) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:441]: ❗ I think this comment means we need to add some logic around release candidacy (and probably pre-release?) to not upload docs for them, if we want parity with the old behavior. presumably we also probably dont want the icon uploaded unless we're uploading to latest, since it's not versioned. 

Though I'm confused why this was intentional in the first place, if it's going to its own filepath, i feel like it should be okay to get its own docs that are loaded up if a user is using that release candidate version

Comment (erohmensing) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:441]: Or maybe since these files are going somwhere else it doesn't matter? I am a little lost on this bc im out of the loop with release candidate stuff, so i'll let @alafanechere chime in here

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (erohmensing) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:135]:
```suggestion
    """Check if the list of uploaded files contains any of the provided keys."""
```
This one broke my brain a little the first time I read it 😆

File: airbyte-ci/connectors/metadata_service/lib/tests/test_gcs_upload.py
Comment (erohmensing) [inline airbyte-ci/connectors/metadata_service/lib/tests/test_gcs_upload.py:403]: I wonder if its worth pulling this case out like the prerelease one, so we can nail down the exact behavior we want in this case (see: other comment about release candidate uploads)

**Review by erohmensing — APPROVED** [review]
Most comments are stylistic. The only thing I'm confused/concerned about is what exactly we want to upload for release candidates, but I'll approve from my side and let Augustin make the call on that one, cause I don't have the answer :D 

@bnchrch One question: what's the plan for getting these all up the first time, i.e. backfilling, I guess? Are we going to backfill or just wait for publishes to kick them up one by one? (I'm still having issues with the dagster backfill, so if we have plans for kicking off a bunch of publishes that update metadata files by adding SHAs to them, that might end up solving both problems...)

Thanks so much for getting this over the line!

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/commands.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/commands.py:51]: Nope! just for ourselves

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:177]: We should! Im just a bit short on time 😅 

Mind if I do that in another PR?

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:240]: Updated

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:446]: The tests should cover this more explicitly now

**Comment by bnchrch** [comment]
@erohmensing heads up that this is back up for review!


Just ran the updated pipeline locally for a manifest + components connector

https://console.cloud.google.com/storage/browser/dev-airbyte-cloud-connector-metadata-service/metadata/airbyte/source-the-guardian-api/0.2.0-dev.7b20c3b7a0?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&cloudshell=false&project=prod-ab-cloud-proj

**Comment by bnchrch** [comment]
Ill merge when im on in the morning and we can start backfilling

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (erohmensing) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:177]: Yeah, no worries!

File: airbyte-ci/connectors/pipelines/README.md
Comment (erohmensing) [reply airbyte-ci/connectors/pipelines/README.md:846]: Oh, true! My bad

**Comment by erohmensing** [comment]
@bnchrch awesome, sounds good! Happy to pair on backfilling, I can do metadata while you do manifests

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/commands.py
Comment (natikgadzhi) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/commands.py:17]: I think this has something to do with the fact that we're executing in Dagster, but why are we using `click.secho` here? /cc @erohmensing (I know the change here is just formatting, too, so question out of curiosity.)

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/commands.py
Comment (natikgadzhi) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/commands.py:51]: Can someone remind me how this change is related to the PR main goal?

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py
Comment (natikgadzhi) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py:43]: I assume it wasn't used?

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/commands.py
Comment (erohmensing) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/commands.py:17]: WDYM? Do you mean as compared to plain `echo`? Just for styling: https://click.palletsprojects.com/en/8.1.x/utils/#ansi-colors

**Pull request merged**
